src/VisualStudio/LiveShare/Impl/GotoDefinitionWithFarHandler.cs

```diff
@@ -18,19 +18,15 @@
 namespace Microsoft.VisualStudio.LanguageServices.LiveShare
 {
     /// <summary>
-    /// Handler for a goto defintion request. This can return files outside the shared
-    /// folder cone. The HostProtocolConverter handles converting files outside the shared
-    /// cone to the vslsexternal scheme.
-    /// This lives in external access because it has a dependency on FAR, which requires the UI thread.
+    /// Typescript uses FAR for goto def as they do not implement the definition service.
+    /// So implement goto def for typescript using FAR.
+    /// FAR has a UI thread dependency.
     /// </summary>
-    internal class GotoDefinitionHandler : ILspRequestHandler<LSP.TextDocumentPositionParams, object, Solution>
+    internal class GotoDefinitionWithFarHandler : ILspRequestHandler<LSP.TextDocumentPositionParams, object, Solution>
```

Comment:
what does `WithFarHandler` mean?

Reply:
Typescript doesn't support the definition service, so instead the approach they took was to use the find usages service, aka GotoDef with far

Reply:
That seems strange... i definitely implemented goto-def for TypeScript.  It was one of the first features i did that i sat on top of roslyn.

Also, if it's using 'Find References', does that mean that if you do 'goto def' with TS, you literally get all references as well?

Reply:
> That seems strange... i definitely implemented goto-def for TypeScript. It was one of the first features i did that i sat on top of roslyn.
TS does implement gotodef, just not the `IGoToDefinitionService` to return definitions, correct @daytonellwanger ?

> Also, if it's using 'Find References', does that mean that if you do 'goto def' with TS, you literally get all references as well?

No, it just uses the find usages service behind the scenes instead of the `IGoToDefinitionService`

Reply:
> No, it just uses the find usages service behind the scenes instead of the 

I'm a little confused by that as well.  This is the entire IFindUsagesService:

```c#
    internal interface IFindUsagesService : ILanguageService
    {
        Task FindReferencesAsync(Document document, int position, IFindUsagesContext context);
        Task FindImplementationsAsync(Document document, int position, IFindUsagesContext context);
    }
```

How does it end up using this?

Reply:
See https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/LiveShare/Impl/GoToDefintionWithFarHandler.cs#L97

Reply:
Interesting...

Still not clear why it was done this way.  TS def supported Go-to-def from teh beginning.  It's massively cheaper as a feature.  Goto-def in TS is basically almost free once semantic info is computed.  Whereas find-refs actually has to 'find' the results by searching all the documents in the project.  :-/

Reply:
We used to use `definitionService.FindDefinitionsAsync`, but that quit returning results around VS 16.1... so we switched to this less efficient implementation. If that gets resolved, we should certainly switch back.

Reply:
> We used to use definitionService.FindDefinitionsAsync, but that quit returning results around VS 16.1... so we switched to this less efficient implementation.

Uh... it def sounds like that should be investigated :)

Reply:
Had a quick chat with @amcasey - TS had an issue with duplicate results for their peek definition, so they changed their IGoToDefinitionService to return empty results for `FindDefinitionsAsync`.  They still implement the service, but only for `TryGoToDefinition`.  

So for now I will leave this how it is, and have filed https://github.com/dotnet/roslyn/issues/38483 to follow up as it requires some extra thought on how to resolve this.

Reply:
@amcasey can you shed some light here? I'm not understanding why this woudl be a problem for TS, but not VB/C#.  Also, it feels *really* unfortunate that we went this direction instead of addressing any underlying issues with IGoToDefinitionService.  it feels like we got some hacks introduced then, and we're not adding more and more around that instead of fixing root cause stuff.

Reply:
May I request that this discussion be held on #38483 ?  We're not adding to this hack here (it already existed unfortunately), just moving it to limit the impact to the typescript liveshare handler + cleaning up the c#/vb implementation to just be a shim.